_posts/2021-02-02-getting-started-amqp.adoc

```diff
@@ -0,0 +1,335 @@
+
+Remember, the movie is encoded into JSON, so we need to help the connector produce a `Movie` from the received JSON.
+
+Create the `org.acme.JsonToObjectConverter` class with the following content:
+
+[source, java]
+----
+package org.acme;
+
+import io.smallrye.reactive.messaging.MessageConverter;
+import io.smallrye.reactive.messaging.amqp.IncomingAmqpMetadata;
+import io.vertx.core.json.JsonObject;
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+import javax.enterprise.context.ApplicationScoped;
+import java.lang.reflect.Type;
+
+@ApplicationScoped
+public class JsonToObjectConverter implements MessageConverter {
+
+    @Override
+    public boolean canConvert(Message<?> in, Type target) {
+        return in.getMetadata(IncomingAmqpMetadata.class)
+                .map(meta -> meta.getContentType().equals("application/json")  && target instanceof Class)
+                .orElse(false);
+
+    }
+
+    @Override
+    public Message<?> convert(Message<?> in, Type target) {
+        return in.withPayload(((JsonObject) in.getPayload()).mapTo((Class<?>) target));
+    }
+}
+----
+
+This class is a _converter_.
+It maps the content of a `Message` to another type.
+In the `canConvert` method, we verify that the incoming message is coming from AMQP (so contain the `IncomingAmqpMetadata` metadata) and that the content-type is set to `application/json`.
+The `convert` method maps the received `JsonObject` into the target type (`Movie` in our case).
+
+With this converter, our `consume` method will receive the `Movie` objects.
+
+== Step 6 - Sending movies from a REST endpoint
+
+It's quite common to send messages to AMQP from a REST endpoint.
+To do this, create the `org.acme.MovieResource` class with the following content:
+
+[source, java]
+----
+package org.acme;
+
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+@Path("/")
+@Consumes(MediaType.APPLICATION_JSON)
+@Produces(MediaType.APPLICATION_JSON)
+public class MovieResource {
+
+    @Inject
+    MovieProducer producer;
+
+    @POST
+    public Response send(Movie movie) {
+        producer.send(movie);
+        // Return an 202 - Accepted response.
+        return Response.accepted().build();
+    }
+}
+----
+
+This class uses the `MovieProducer` we implemented above to send the `movies`.
+You could also use an `Emitter` directly.
+
+== Step 7 - Let's get this running!
+
+Well, first, we need an AMQP broker.
+You can follow the http://activemq.apache.org/components/artemis/documentation/1.0.0/running-server.html[Getting Started with Artemis] documentation, or use the following `docker-compose.yaml` file:
+
+[source, yaml]
+----
+version: '2'
+
+services:
+
+  artemis:
+    image: vromero/activemq-artemis:2-alpine-latest
+    ports:
+      - "8161:8161"
+      - "61616:61616"
+      - "5672:5672"
```

Comment:
Nitpicking but I would put 5672 first since its the AMQP port, lowest numerically, and only one actually used by the example.

Reply:
Fixed